clean_code/algorithms/creation_algorithms.py

```diff
@@ -0,0 +1,66 @@
+from math import factorial
+
+"""This module supplies fast algorithms for pattern-related creations:
+   all permutations of a staring, duplicate-deletion."""
+
+def create_trimmed_string(base_string):
+    """deletes duplicates in a string. example: aaabbaac -> abac"""
+    trimmed_elements = []
+    trimmed_elements.append(base_string[:1])
+    for index in range(1, len(base_string)):
+        if base_string[index] != base_string[index-1]:
+            trimmed_elements.append(base_string[index])
+    return ''.join(trimmed_elements)
+
+
+def generate_permutations_iterably(elements):
+    """returns all the permutations of a string. the function is non-recursive"""
+    permutations = set()
+    final_set = set()
+    expectd_permutations_number = factorial(len(elements))
+
+    new_combinations = _first_cell_permutations(elements)
+    permutations.update(new_combinations)
+
+    while len(final_set) < expectd_permutations_number:
+        permutations_copy = permutations.copy()
+        permutations = set()
+
+        for permutation in permutations_copy:
+            new_combinations = _first_cell_permutations(permutation)
+            final_set.update(new_combinations[1:])
+            permutations.update(new_combinations[1:])
```

Comment:
1. add a comment to explain it. (it seems voodoo)
2. maybe the update of `final_set` can be outside the `for`